### PR TITLE
HADOOP-15729. [s3a] Allow core threads to time out.

### PR DESCRIPTION
**Pull request by mackrorysd**

### CODE REVIEW

**Comment by steveloughran** [comment]
LGTM. Testing?

**Comment by mackrorysd** [comment]
Details of some performance testing of various approaches on the JIRA - dropping core threads had a slightly surprising drop in performance of 1GB uploads, which is why I decided to simply allow timeout instead of messing with the way the threads scale up. Have tested dropping core threads (by tweaking max threads) to a low number on a large cluster to verify it fixes the problem I was seeing. I didn't initially feel this was appropriate for a JUnit test because of the delays involved in ramping up the number of threads, letting them time out, etc. But I made it a scale test and updated the PR.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 59 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1048 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 26 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 738 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | trunk passed |
| 0 | spotbugs | 56 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 54 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 28 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| -0 | checkstyle | 15 | hadoop-tools/hadoop-aws: The patch generated 1 new + 10 unchanged - 0 fixed = 11 total (was 10) |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 703 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 20 | the patch passed |
| +1 | findbugs | 61 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 276 | hadoop-aws in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3290 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1075 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8f4dc3bd1327 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6a3433b |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/1/testReport/ |
| Max. process+thread count | 413 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/scale/ITestS3AConcurrentOps.java [hunk 1]
```diff
@@ -66,20 +65,43 @@ protected int getTestTimeoutSeconds() {
...
+    URI rootURI = new URI(conf.get(TEST_FS_S3A_NAME));
+    s3a.initialize(rootURI, conf);
+    return s3a;
+  }
+
+  @After
+  public void teardown() throws Exception {
+    super.teardown();
+    if (auxFs != null) {
+      auxFs.delete(testRoot, true);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/scale/ITestS3AConcurrentOps.java]: does this need an `auxFs.close()`

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/scale/ITestS3AConcurrentOps.java [hunk 2]
```diff
@@ -89,38 +111,48 @@ public void setup() throws Exception {
...
+                "testParallelRename" + count.getAndIncrement());
+          }
+        });
+    try {
+      ((ThreadPoolExecutor)executor).prestartAllCoreThreads();
+      Future<Boolean>[] futures = new Future[concurrentRenames];
+      for (int i = 0; i < concurrentRenames; i++) {
+        final int index = i;
+        futures[i] = executor.submit(new Callable<Boolean>() {
+          @Override
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/scale/ITestS3AConcurrentOps.java]: you could just do this as a lambda expression; we'd only need the callable for any branch-2 backport

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/scale/ITestS3AConcurrentOps.java [hunk 3]
```diff
@@ -89,38 +111,48 @@ public void setup() throws Exception {
...
+          public Boolean call() throws Exception {
+            NanoTimer timer = new NanoTimer();
+            boolean result = fs.rename(source[index], target[index]);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/scale/ITestS3AConcurrentOps.java]: Have a look at how DurationInfo works/is used; you can hand off the collection and logging to it

**Comment by steveloughran** [comment]
bq. dropping core threads had a slightly surprising drop in performance of 1GB uploads, 

interesting. 

Patch LGTM; a couple of comments on the tests (duration info, lamba vs callable) aren't that important given its an existing test you are tweaking, rather than a new one. The only think I do want to make sure is the auxFs is closed: we don't want to leak threads after all :)

**Comment by steveloughran** [comment]
BTW, I'm +1 pending the auxFs.close call

**Comment by mackrorysd** [comment]
Thanks. Adding an auxFs.close call, and also taking care of closing tinyThreadPoolFs.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 47 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1089 | trunk passed |
| +1 | compile | 33 | trunk passed |
| +1 | checkstyle | 19 | trunk passed |
| +1 | mvnsite | 38 | trunk passed |
| +1 | shadedclient | 677 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | trunk passed |
| 0 | spotbugs | 62 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 59 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 30 | the patch passed |
| +1 | javac | 30 | the patch passed |
| -0 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 1 new + 10 unchanged - 0 fixed = 11 total (was 10) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 688 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 20 | the patch passed |
| +1 | findbugs | 61 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 283 | hadoop-aws in the patch passed. |
| +1 | asflicense | 27 | The patch does not generate ASF License warnings. |
| | | 3287 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1075 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux edd634cc0cb5 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 71c21f0 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/2/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1075/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request merged**